lensless/recon/utils.py

```diff
@@ -129,3 +146,384 @@ def process(image, noise_level):
+            pbar.set_description(f"loss : {mean_loss}")
+            i += 1
+
+        return mean_loss
+
+    def evaluate(self, mean_loss, save_pt):
+        """
+        Evaluate the reconstruction algorithm on the test dataset.
+
+        Parameters
+        ----------
+        mean_loss : float
+            Mean loss of the last epoch.
+        save_pt : str
+            Path to save metrics dictionary to. If None, no logging of metrics.
+        """
+        if self.test_dataset is None:
+            return
+        # benchmarking
+        current_metrics = benchmark(self.recon, self.test_dataset, batchsize=10)
+
+        # update metrics with current metrics
+        self.metrics["LOSS"].append(mean_loss)
+        for key in current_metrics:
+            self.metrics[key].append(current_metrics[key])
+
+        if save_pt:
+            # save dictionary metrics to file with json
+            with open(os.path.join(save_pt, "metrics.json"), "w") as f:
+                json.dump(self.metrics, f)
+
+    def on_epoch_end(self, mean_loss, save_pt):
+        """
+        Called at the end of each epoch.
+
+        Parameters
+        ----------
+        mean_loss : float
+            Mean loss of the last epoch.
+        save_pt : str
+            Path to save metrics dictionary to. If None, no logging of metrics.
+        """
+        if save_pt is None:
+            # Use current directory
+            save_pt = os.getcwd()
+
+        # save model
+        self.save(path=save_pt, include_optimizer=False)
+        self.evaluate(mean_loss, save_pt)
+
+    def train(self, n_epoch=1, save_pt=None, disp=-1):
+        """
+        Train the reconstruction algorithm.
+
+        Parameters
+        ----------
+        n_epoch : int, optional
+            Number of epochs to train for, by default 1
+        save_pt : str, optional
+            Path to save metrics dictionary to. If None, use current directory, by default None
+        disp : int, optional
+            Display interval, if -1, no display. Default is -1.
+        """
+
+        start_time = time.time()
+
+        for epoch in range(n_epoch):
+            print(f"Epoch {epoch} with learning rate {self.scheduler.get_last_lr()}")
+            mean_loss = self.train_epoch(self.train_dataloader, disp=disp)
+            self.on_epoch_end(mean_loss, save_pt)
+            self.scheduler.step()
+
+        print(f"Train time : {time.time() - start_time} s")
+
+    def save(self, path="recon", include_optimizer=False):
+        """
+        Save state of reconstruction algorithm.
+
+        Parameters
+        ----------
+        path : str, optional
+            Path to save model to, by default "recon"
+        include_optimizer : bool, optional
+            Whether to include optimizer state, by default False
+
+        """
+        # create directory if it does not exist
+        if not os.path.exists(path):
+            os.makedirs(path)
+
+        # TODO : ADD mask support
```

Comment:
@YohannPerron what needs to be done for this?

Reply:
Quite a bit of work, it will be done in a separate PR.